File: mail.sharexpress.in/backend/app/api/v1/mail.py
"""FastAPI routes for mail operations."""
from fastapi import APIRouter, Depends, status, Query
from motor.motor_asyncio import AsyncIOMotorDatabase
from typing import List, Optional

from app.core.deps import get_database, get_current_user
from app.models.email import EmailOut, EmailListItem, ComposeRequest, DraftSave, MoveRequest, FlagRequest, SearchRequest
from app.services.mail_service import MailService

router = APIRouter()

@router.get("/folder/{folder_name}", response_model=List[EmailOut])
async def get_folder_emails(
    folder_name: str,
    limit: int = Query(50, ge=1, le=100),
    skip: int = Query(0, ge=0),
    db: AsyncIOMotorDatabase = Depends(get_database),
    current_user: dict = Depends(get_current_user)
):
    """Retrieve emails from a specific folder (inbox, sent, drafts, trash, spam, etc.)"""
    emails = await MailService.get_emails(db, current_user["email"], folder_name, limit, skip)
    return [EmailOut(**e) for e in emails]

@router.get("/detail/{email_id}", response_model=EmailOut)
async def get_email_detail(
    email_id: str,
    db: AsyncIOMotorDatabase = Depends(get_database),
    current_user: dict = Depends(get_current_user)
):
    """Retrieve details of a single email (marks read automatically)."""
    email_doc = await MailService.get_email_by_id(db, email_id, current_user["email"])
    return EmailOut(**email_doc)

@router.post("/compose", response_model=EmailOut, status_code=status.HTTP_201_CREATED)
async def compose_email(
    req: ComposeRequest,
    db: AsyncIOMotorDatabase = Depends(get_database),
    current_user: dict = Depends(get_current_user)
):
    """Send an email immediately (SMTP + Sent box storage)."""
    email_doc = await MailService.send_email_flow(db, current_user["email"], req)
    return EmailOut(**email_doc)

@router.post("/draft", response_model=EmailOut, status_code=status.HTTP_201_CREATED)
async def save_draft(
    draft: DraftSave,
    db: AsyncIOMotorDatabase = Depends(get_database),
    current_user: dict = Depends(get_current_user)
):
    """Save an email draft without sending."""
    draft_doc = await MailService.create_draft(db, current_user["email"], draft)
    return EmailOut(**draft_doc)

@router.post("/move", status_code=status.HTTP_200_OK)
async def move_emails(
    req: MoveRequest,
    db: AsyncIOMotorDatabase = Depends(get_database),
    current_user: dict = Depends(get_current_user)
):
    """Move list of emails to a target folder (e.g. archive, trash)."""
    modified = await MailService.move_emails(db, current_user["email"], req.email_ids, req.target_folder)
    return {"modified_count": modified, "status": "success"}

@router.post("/flag", status_code=status.HTTP_200_OK)
async def flag_emails(
    req: FlagRequest,
    db: AsyncIOMotorDatabase = Depends(get_database),
    current_user: dict = Depends(get_current_user)
):
    """Mark list of emails as read/unread or starred/unstarred."""
    modified = await MailService.flag_emails(
        db, current_user["email"], req.email_ids, read=req.read, starred=req.starred
    )
    return {"modified_count": modified, "status": "success"}

@router.delete("/purge", status_code=status.HTTP_200_OK)
async def purge_emails(
    email_ids: List[str] = Query(...),
    db: AsyncIOMotorDatabase = Depends(get_database),
    current_user: dict = Depends(get_current_user)
):
    """Permanently delete specified emails (bypass Trash)."""
    deleted = await MailService.delete_permanently(db, current_user["email"], email_ids)
    return {"deleted_count": deleted, "status": "success"}

File: mail.sharexpress.in/frontend/src/api/mail.js
import apiClient from './axios';

export const mailApi = {
  getFolderEmails: async (folder, limit = 50, skip = 0) => {
    const response = await apiClient.get(`/mail/folder/${folder}`, {
      params: { limit, skip }
    });
    return response.data;
  },

  getEmailDetail: async (id) => {
    const response = await apiClient.get(`/mail/detail/${id}`);
    return response.data;
  },

  sendEmail: async (data) => {
    const response = await apiClient.post('/mail/compose', data);
    return response.data;
  },

  saveDraft: async (data) => {
    const response = await apiClient.post('/mail/draft', data);
    return response.data;
  },

  moveEmails: async (emailIds, targetFolder) => {
    const response = await apiClient.post('/mail/move', {
      email_ids: emailIds,
      target_folder: targetFolder
    });
    return response.data;
  },

  flagEmails: async (emailIds, flags) => {
    const response = await apiClient.post('/mail/flag', {
      email_ids: emailIds,
      ...flags
    });
    return response.data;
  },

  purgeEmails: async (emailIds) => {
    const response = await apiClient.delete('/mail/purge', {
      params: { email_ids: emailIds }
    });
    return response.data;
  },

  uploadAttachment: async (file) => {
    const formData = new FormData();
    formData.append('file', file);
    const response = await apiClient.post('/mail/attachment/upload', formData, {
      headers: {
        'Content-Type': 'multipart/form-data'
      }
    });
    return response.data;
  },

  downloadAttachment: async (url) => {
    const response = await apiClient.get(url, {
      responseType: 'blob'
    });
    return response;
  },

  // Admin APIs
  adminListUsers: async () => {
    const response = await apiClient.get('/admin/users');
    return response.data;
  },

  adminCreateUser: async (userData) => {
    const response = await apiClient.post('/admin/users', userData);
    return response.data;
  },

  adminUpdateUser: async (id, userData) => {
    const response = await apiClient.put(`/admin/users/${id}`, userData);
    return response.data;
  },

  adminDeleteUser: async (id) => {
    await apiClient.delete(`/admin/users/${id}`);
  }
};

export default mailApi;
